README.md
Color wand ✨ 🍭🧙‍♂️ [![Build Status](https://travis-ci.com/tio-design/color_wand.svg?branch=feat%2Ftravis)](https://travis-ci.com/tio-design/color_wand)
=====
A color library which can be used to create, convert, alter, and compare colors to be used in both Flutter and the web.

Installation 💻
-----
1. Add this to your package's `pubspec.yaml` file:

```yaml
dependencies:
  color: any
```

2. Get the package using your IDE's GUI or via command line with

```bash
$ pub get
```

3. Import the `color_wand.dart` file in your app

```dart
import 'package:color_wand/color_wand.dart';
```

Usage 🥒
-----
Color objects can be constructed by either using the factory constructors of the `Color` interface or by directly 
instantiating an `RgbColor` or an `HslColor` using one of their constructors.

- From HEX: `Color.hex(0xRRGGBBAA)`
- From a HEX string: `Color.parseHex('#RRGGBBAA')`
- From RGB: `Color.rgb(red: red, green: green, blue: blue)`
- From RGBA: `Color.rgba(red: red, green: green, blue: blue, alpha: alpha)`
- From RGBO: `Color.rgbo(red: red, green: green, blue: blue, opacity: opacity)`
- From HSL: `Color.hsl(hue: hue, saturation: saturation, lightness: lightness)`
- From HSLO: `Color.hslo(hue: hue, saturation: saturation, lightnesss: lightness, opacity: opacity)`
- From the lower 32 bits of an integer to be used e.g. for fast conversion to and from the Flutter color lib: `RgbColor(value)`

Colors are immutable, and most of them can be created using const constructors:
```dart
Color color = const Color.hex(0x12345678);
```

Colors can be converted from one color space to another by calling `toRgb` or `toHsl` on them.

Colors can be compared using the `==` operator, which will implicitly convert the color to an `RgbColor` and then compare there values.

Colors can be converted to a CSS color string by using one of the different implementations of the `CssColorFormat` interface.

Available formats are:
- `RgbFormat`: `rgb(red, green, blue)` or `rgb(red, green, blue, alpha)`
- `RgbaFormat`: `rgba(red, green, blue)` or `rgba(red, green, blue, alpha)`
- `HslFormat`: `hsl(hue, saturation, lightness)` or `hsl(hue, saturation, lightness, alpha)`
- `HslaFormat`: `hsla(hue, saturation, lightness)` or `hsla(hue, saturation, lightness, alpha)`
- `HexFormat` `#RRGGBB`or `#RRGGBBAA`

Most of the formats can be customized to e.g. represent the values as percentages or decimal numbers.

```dart
final rgbFormat = RgbFormat(rgbStyle: PercentOrAlpha.alpha);
final hslFormat = HslFormat(hueStyle: DegreeOrDecimal.degree);
final color = Color.hsl(24, 100, 50);
assert(rgbFormat.format(color) == 'rgb(255, 102, 0)');
assert(hslFormat.format(color) == 'hsl(24deg, 100%, 50%)');
```

changelog
0.0.1
-----
* Initial version

0.0.2
-----
* Drop version of meta to 1.0.0 because of compatibility issues with other packages

0.0.3
-----
* Reformat with dartfmt

0.0.4
-----
* Fix issue where HexFormat would cut of the red part of the hex string when red was 0


test/color_test.dart
import 'package:color_wand/src/color.dart';
import 'package:color_wand/src/hsl_color.dart';
import 'package:color_wand/src/rgb_color.dart';
import 'package:test/test.dart';

void main() {
  group("Should construct a color", () {
    test("when using the Color.hex factory.", () {
      final color = Color.hex(0x12345678);
      expect(color, TypeMatcher<RgbColor>());
      expect(color.toRgb().value, 0x78123456);
    });
    test("when using the Color.rgb factory.", () {
      final color = Color.rgb(red: 120, green: 121, blue: 122);
      final rgbColor = color.toRgb();
      expect(color, TypeMatcher<RgbColor>());
      expect(rgbColor.red, 120);
      expect(rgbColor.green, 121);
      expect(rgbColor.blue, 122);
    });
    test("when using the Color.rgba factory.", () {
      final color = Color.rgba(red: 120, green: 121, blue: 122, alpha: 123);
      final rgbColor = color.toRgb();
      expect(color, TypeMatcher<RgbColor>());
      expect(rgbColor.red, 120);
      expect(rgbColor.green, 121);
      expect(rgbColor.blue, 122);
      expect(rgbColor.alpha, 123);
    });
    test("when using the Color.rgbo factory.", () {
      final color = Color.rgbo(red: 120, green: 121, blue: 122, opacity: 0.5);
      final rgbColor = color.toRgb();
      expect(color, TypeMatcher<RgbColor>());
      expect(rgbColor.red, 120);
      expect(rgbColor.green, 121);
      expect(rgbColor.blue, 122);
      expect(rgbColor.opacity, 0.5);
    });
    test("when using the Color.hsl factory.", () {
      final color = Color.hsl(hue: 10, saturation: 11, lightness: 12);
      final hslColor = color.toHsl();
      expect(color, TypeMatcher<HslColor>());
      expect(hslColor.hue, 10);
      expect(hslColor.saturation, 11);
      expect(hslColor.lightness, 12);
    });
    test("when using the Color.hslo factory.", () {
      final color =
          Color.hslo(hue: 10, saturation: 11, lightness: 12, opacity: 0.5);
      final hslColor = color.toHsl();
      expect(color, TypeMatcher<HslColor>());
      expect(hslColor.hue, 10);
      expect(hslColor.saturation, 11);
      expect(hslColor.lightness, 12);
      expect(hslColor.opacity, 0.5);
    });
    test("when using the parseHex function with 8 bytes supplied.", () {
      final color = Color.parseHex('#12345678').toRgb();
      expect(color.value, 0x78123456);
    });
    test("when using the parseHex function with only 6 bytes supplied.", () {
      final color = Color.parseHex('#345678').toRgb();
      expect(color.value, 0xFF345678);
    });
    test("when using the parseHex function withouth a hash prefix.", () {
      final color = Color.parseHex('12345678').toRgb();
      expect(color.value, 0x78123456);
    });
    test(
        "with the value brought into range when using the parseHex function with a hex longer than 8 bytes supplied.",
        () {
      final color = Color.parseHex('#9A34567812').toRgb();
      expect(color.value, 0x12345678);
    });
  });

  group("Should throw an error when constructing a color", () {
    test("using the parseHex function with an invalid hext string", () {
      final createInvalidColor = () => Color.parseHex('invalid');
      expect(createInvalidColor, throwsException);
    });
  });

  group("Should return the correct value when getting the", () {
    test("isOpaque value of a transparent Color.", () {
      final color =
          Color.hslo(hue: 10, saturation: 11, lightness: 12, opacity: 0.5);
      expect(color.isOpaque, false);
    });
    test("isOpaque value of an opaque Color.", () {
      final color =
          Color.hslo(hue: 10, saturation: 11, lightness: 12, opacity: 1);
      expect(color.isOpaque, true);
    });
    test("isTransparent value of a transparent Color.", () {
      final color =
          Color.hslo(hue: 10, saturation: 11, lightness: 12, opacity: 0.5);
      expect(color.isTransparent, true);
    });
    test("isTransparent value of an opaque Color.", () {
      final color =
          Color.hslo(hue: 10, saturation: 11, lightness: 12, opacity: 1);
      expect(color.isTransparent, false);
    });
  });

  group("Should compare different instances", () {
    test("of the same color in different color spaces correctly", () {
      final instance1 = HslColor.fromHslo(
          hue: 100, saturation: 80, lightness: 50, opacity: 0.5);
      final instance2 = RgbColor.fromHex(0x5de61980);

      expect(instance1 == instance2, true);
    });
    test("of different colors in different color spaces correctly", () {
      final instance1 = HslColor.fromHslo(
          hue: 100, saturation: 20, lightness: 50, opacity: 0.5);
      final instance2 = RgbColor.fromHex(0x5de41980);

      expect(instance1 == instance2, false);
    });
  });

  group("Should calculate the relative luminance correctly when", () {
    test("using the isDark getter on a dark color.", () {
      final darkColor = Color.hex(0x000000);
      final isDark = darkColor.isDark;
      expect(isDark, true);
    });
    test("using the isDark getter on a bright color.", () {
      final brightColor = Color.hex(0xFFFFFF);
      final isDark = brightColor.isDark;
      expect(isDark, false);
    });
    test("using the isBright getter on a dark color.", () {
      final darkColor = Color.hex(0x000000);
      final isBright = darkColor.isBright;
      expect(isBright, false);
    });
    test("using the isBright getter on a bright color.", () {
      final brightColor = Color.hex(0xFFFFFF);
      final isBright = brightColor.isBright;
      expect(isBright, true);
    });
  });
}


test/hsl_color_test.dart
import 'package:color_wand/src/color.dart';
import 'package:color_wand/src/hsl_color.dart';
import 'package:color_wand/src/rgb_color.dart';
import 'package:test/test.dart';

void main() {
  group("Should construct a HslColor", () {
    test(
        "with max opacity when using the HslColor.fromHslo constructor with no opacity supplied",
        () {
      final color = HslColor.fromHslo(hue: 100, saturation: 80, lightness: 50);
      expect(color.opacity, Color.maxOpacity);
      expect(color.hue, 100);
      expect(color.saturation, 80);
      expect(color.lightness, 50);
    });
    test(
        "with opacity when using the HslColor.fromHslo constructor with opacity supplied",
        () {
      final color = HslColor.fromHslo(
          hue: 100, saturation: 80, lightness: 50, opacity: 0.5);
      expect(color.opacity, 0.5);
      expect(color.hue, 100);
      expect(color.saturation, 80);
      expect(color.lightness, 50);
    });
  });

  group("Should return a new manipulated instance", () {
    test("with the changed opacity when changing the opacity.", () {
      final color = HslColor.fromHslo(
          hue: 100, saturation: 80, lightness: 50, opacity: 0.5);
      final newColor = color.withOpacity(0.25);
      expect(newColor.opacity, 0.25);
    });
    test("with the changed opacity when changing the opacity.", () {
      final color = HslColor.fromHslo(
          hue: 100, saturation: 80, lightness: 50, opacity: 0.5);
      final newColor = color.withOpacity(0.25);
      expect(newColor.opacity, 0.25);
    });
    test("with the changed lightness when lightening by steps.", () {
      final color = HslColor.fromHslo(hue: 100, saturation: 80, lightness: 50);
      final newColor = color.lightenBySteps(10);
      expect(newColor.lightness, color.lightness + 10);
    });
    test(
        "with the changed lightness brought into range when lightening to an out of bounds value by steps.",
        () {
      final color = HslColor.fromHslo(hue: 100, saturation: 80, lightness: 50);
      final newColor = color.lightenBySteps(60);
      expect(newColor.lightness, 10);
    });
    test("with the changed lightness when lightening by fraction.", () {
      final color = HslColor.fromHslo(hue: 100, saturation: 80, lightness: 50);
      final newColor = color.lightenByFraction(0.5);
      expect(newColor.lightness, 75);
    });
    test(
        "with the changed lightness brought into range when lightening to an out of bounds value by fraction.",
        () {
      final color = HslColor.fromHslo(hue: 100, saturation: 80, lightness: 50);
      final newColor = color.lightenByFraction(2);
      expect(newColor.lightness, 50);
    });
    test("with the changed lightness when setting the lightness.", () {
      final color = HslColor.fromHslo(hue: 100, saturation: 80, lightness: 50);
      final newColor = color.withLightness(60);
      expect(newColor.lightness, 60);
    });
    test(
        "with the changed lightness brought into range when setting the lightness to an out of bounds value.",
        () {
      final color = HslColor.fromHslo(hue: 100, saturation: 80, lightness: 50);
      final newColor = color.withLightness(110);
      expect(newColor.lightness, 10);
    });
    test("with the changed lightness when darkening by steps.", () {
      final color = HslColor.fromHslo(hue: 100, saturation: 80, lightness: 50);
      final newColor = color.darkenBySteps(10);
      expect(newColor.lightness, color.lightness - 10);
    });
    test(
        "with the changed lightness brought into range when darkening to an out of bounds value by steps.",
        () {
      final color = HslColor.fromHslo(hue: 100, saturation: 80, lightness: 50);
      final newColor = color.darkenBySteps(60);
      expect(newColor.lightness, 90);
    });
    test("with the changed lightness when darkening by fraction.", () {
      final color = HslColor.fromHslo(hue: 100, saturation: 80, lightness: 50);
      final newColor = color.darkenByFraction(0.5);
      expect(newColor.lightness, color.lightness / 2);
    });
    test(
        "with the changed lightness brought into range when darkening to an out of bounds value by steps.",
        () {
      final color = HslColor.fromHslo(hue: 100, saturation: 80, lightness: 50);
      final newColor = color.darkenByFraction(1.5);
      expect(newColor.lightness, 75);
    });
  });

  group("Should return a converted instance", () {
    test(
        "of RgbColor when using the to Rgb conversion function on an HslColor.",
        () {
      final color = HslColor.fromHslo(
          hue: 100, saturation: 80, lightness: 50, opacity: 0.5);
      final rgbColor = color.toRgb();
      expect(rgbColor, TypeMatcher<RgbColor>());
      expect(rgbColor.opacity, color.opacity);
      expect(rgbColor.red, 93);
      expect(rgbColor.green, 230);
      expect(rgbColor.blue, 25);
    });
    test(
        "of HslColor when unsing the to Hsl conversion function on an HslColor.",
        () {
      final color = HslColor.fromHslo(
          hue: 100, saturation: 80, lightness: 50, opacity: 0.5);
      final hslColor = color.toHsl();
      expect(hslColor, TypeMatcher<HslColor>());
      expect(hslColor, equals(color));
    });
  });

  group("Should return the correct value when getting the", () {
    test("opacity of an HslColor.", () {
      final color = HslColor.fromHslo(
          hue: 100, saturation: 80, lightness: 50, opacity: 0.5);
      expect(color.opacity, 0.5);
    });
    test("hue value of an HslColor.", () {
      final color = HslColor.fromHslo(
          hue: 100, saturation: 80, lightness: 50, opacity: 0.5);
      expect(color.hue, 100);
    });
    test("saturation of an HslColor.", () {
      final color = HslColor.fromHslo(
          hue: 100, saturation: 80, lightness: 50, opacity: 0.5);
      expect(color.saturation, 80);
    });
    test("lightness of an HslColor.", () {
      final color = HslColor.fromHslo(
          hue: 100, saturation: 80, lightness: 50, opacity: 0.5);
      expect(color.lightness, 50);
    });
  });

  group("Should compare different instances", () {
    test("of the same color correctly", () {
      final instance1 = HslColor.fromHslo(
          hue: 100, saturation: 80, lightness: 50, opacity: 0.5);
      final instance2 = HslColor.fromHslo(
          hue: 100, saturation: 80, lightness: 50, opacity: 0.5);

      expect(instance1 == instance2, true);
    });
    test("of different colors correctly", () {
      final instance1 = HslColor.fromHslo(
          hue: 100, saturation: 80, lightness: 50, opacity: 0.5);
      final instance2 = HslColor.fromHslo(
          hue: 101, saturation: 81, lightness: 51, opacity: 0.51);

      expect(instance1 == instance2, false);
    });
  });
}


test/rgb_color_test.dart
import 'package:color_wand/src/hsl_color.dart';
import 'package:color_wand/src/rgb_color.dart';
import 'package:test/test.dart';

void main() {
  group("Should construct a RgbColor", () {
    test(
        "with min alpha when using the RgbColor constructor with only 6 bytes supplied.",
        () {
      final color = RgbColor(0x345678);
      expect(color.alpha, RgbColor.minAlpha);
      expect(color.red, 52);
      expect(color.green, 86);
      expect(color.blue, 120);
    });
    test(
        "with alpha when using the RgbColor constructor with 8 bytes supplied.",
        () {
      final color = RgbColor(0x12345678);
      expect(color.alpha, 18);
      expect(color.red, 52);
      expect(color.green, 86);
      expect(color.blue, 120);
    });
    test(
        "with the value brought into range when using the RgbColor constructor with more than 8 bytes supplied.",
        () {
      final color = RgbColor.fromHex(0x9A34567812);
      expect(color.alpha, 18);
      expect(color.red, 52);
      expect(color.green, 86);
      expect(color.blue, 120);
    });
    test(
        "with max alpha when using the RgbColor.fromHex constructor with only 6 bytes supplied.",
        () {
      final color = RgbColor.fromHex(0x345678);
      expect(color.alpha, RgbColor.maxAlpha);
      expect(color.red, 52);
      expect(color.green, 86);
      expect(color.blue, 120);
    });
    test(
        "with alpha when using the RgbColor.fromHex constructor with 8 bytes supplied.",
        () {
      final color = RgbColor.fromHex(0x34567812);
      expect(color.alpha, 18);
      expect(color.red, 52);
      expect(color.green, 86);
      expect(color.blue, 120);
    });
    test(
        "with the hex value brought into range when using the RgbColor.fromHex constructor with an out of range hex value.",
        () {
      final color = RgbColor.fromHex(0x9A34567812);
      expect(color.alpha, 18);
      expect(color.red, 52);
      expect(color.green, 86);
      expect(color.blue, 120);
    });
    test(
        "with max alpha when using the RgbColor.fromRgba constructor with no alpha supplied.",
        () {
      final color = RgbColor.fromRgba(red: 52, green: 86, blue: 120);
      expect(color.alpha, RgbColor.maxAlpha);
      expect(color.red, 52);
      expect(color.green, 86);
      expect(color.blue, 120);
    });
    test(
        "with alpha when using the RgbColor.fromRgba constructor with alpha supplied.",
        () {
      final color = RgbColor.fromRgba(red: 52, green: 86, blue: 120, alpha: 12);
      expect(color.alpha, 12);
      expect(color.red, 52);
      expect(color.green, 86);
      expect(color.blue, 120);
    });
    test(
        "with the alpha value brought into range when using the RgbColor.fromRgba constructor with an out of range alpha value.",
        () {
      final color = RgbColor.fromRgba(red: 0, green: 0, blue: 0, alpha: 257);
      expect(color.alpha, 1);
    });
    test(
        "with the red value brought into range when using the RgbColor.fromRgba constructor with an out of range red value.",
        () {
      final color = RgbColor.fromRgba(red: 257, green: 0, blue: 0);
      expect(color.red, 1);
    });
    test(
        "with the green value brought into range when using the RgbColor.fromRgba constructor with an out of range green value.",
        () {
      final color = RgbColor.fromRgba(red: 0, green: 257, blue: 0);
      expect(color.green, 1);
    });
    test(
        "with the blue value brought into range when using the RgbColor.fromRgba constructor with an out of range blue value.",
        () {
      final color = RgbColor.fromRgba(red: 0, green: 0, blue: 257);
      expect(color.blue, 1);
    });
    test(
        "with max alpha when using the RgbColor.fromRgbo constructor with no opacity supplied.",
        () {
      final color = RgbColor.fromRgbo(red: 52, green: 86, blue: 120);
      expect(color.alpha, RgbColor.maxAlpha);
      expect(color.red, 52);
      expect(color.green, 86);
      expect(color.blue, 120);
    });
    test(
        "with alpha when using the RgbColor.fromRgbo constructor with opacity supplied.",
        () {
      final color =
          RgbColor.fromRgbo(red: 52, green: 86, blue: 120, opacity: 0.05);
      expect(color.alpha, 13);
      expect(color.red, 52);
      expect(color.green, 86);
      expect(color.blue, 120);
    });
    test(
        "with the opacity value brought into range when using the RgbColor.fromRgbo constructor with an out of range opacity value.",
        () {
      final color = RgbColor.fromRgbo(red: 0, green: 0, blue: 0, opacity: 1.2);
      expect(color.opacity, 0.2);
    });
    test(
        "with the red value brought into range when using the RgbColor.fromRgbo constructor with an out of range red value.",
        () {
      final color = RgbColor.fromRgbo(red: 257, green: 0, blue: 0);
      expect(color.red, 1);
    });
    test(
        "with the green value brought into range when using the RgbColor.fromRgbo constructor with an out of range green value.",
        () {
      final color = RgbColor.fromRgbo(red: 0, green: 257, blue: 0);
      expect(color.green, 1);
    });
    test(
        "with the blue value brought into range when using the RgbColor.fromRgbo constructor with an out of range blue value.",
        () {
      final color = RgbColor.fromRgbo(red: 0, green: 0, blue: 257);
      expect(color.blue, 1);
    });
  });

  group("Should return a new manipulated instance", () {
    test("with the changed alpha when changing the alpha.", () {
      final color = RgbColor.fromHex(0x123456);
      final newColor = color.withAlpha(120).toRgb();
      expect(newColor.alpha, 120);
      expect(newColor.value, 0x78123456);
    });
    test(
        "with the changed opacity and rounded up alpha when changing the opacity.",
        () {
      final color = RgbColor.fromHex(0x123456);
      final newColor = color.withOpacity(0.47).toRgb();
      expect(newColor.opacity, 0.47);
      expect(newColor.value, 0x78123456);
    });
    test(
        "with the changed opacity and rounded down alpha when changing the opacity.",
        () {
      final color = RgbColor.fromHex(0x123456);
      final newColor = color.withOpacity(0.48).toRgb();
      expect(newColor.opacity, 0.48);
      expect(newColor.value, 0x7a123456);
    });
  });

  group("Should return a converted instance", () {
    test(
        "of HslColor when using the to Hsl conversion function on an RgbColor.",
        () {
      final color = RgbColor.fromHex(0x123456).withOpacity(0.5);
      final hslColor = color.toHsl();
      expect(hslColor, TypeMatcher<HslColor>());
      expect(hslColor.opacity, color.opacity);
      expect(hslColor.hue, 210);
      expect(hslColor.saturation, 65.38);
      expect(hslColor.lightness, 20.39);
    });
    test(
        "of RgbColor when unsing the to Rgb conversion function on an RgbColor.",
        () {
      final color = RgbColor.fromHex(0x123456).withOpacity(0.5);
      final rgbColor = color.toRgb();
      expect(rgbColor, TypeMatcher<RgbColor>());
      expect(rgbColor, equals(color));
    });
  });

  group("Should return the correct value when getting the", () {
    test("opacity of an RgbColor.", () {
      final color = RgbColor.fromHex(0x123456).withOpacity(0.5);
      expect(color.opacity, 0.5);
    });
    test("alpha value of an RgbColor.", () {
      final color = RgbColor.fromHex(0x12345678).toRgb();
      expect(color.alpha, 120);
    });
    test("red value of an RgbColor.", () {
      final color = RgbColor.fromHex(0x123456).toRgb();
      expect(color.red, 18);
    });
    test("alpha value of an RgbColor.", () {
      final color = RgbColor.fromHex(0x123456).toRgb();
      expect(color.green, 52);
    });
    test("alpha value of an RgbColor.", () {
      final color = RgbColor.fromHex(0x123456).toRgb();
      expect(color.blue, 86);
    });
  });

  group("Should compare different instances", () {
    test("of the same color correctly", () {
      final instance1 = RgbColor.fromHex(0x123456).withOpacity(0.5);
      final instance2 = RgbColor.fromHex(0x123456).withOpacity(0.5);

      expect(instance1 == instance2, true);
    });
    test("of different colors correctly", () {
      final instance1 = RgbColor.fromHex(0x123456).withOpacity(0.5);
      final instance2 = RgbColor.fromHex(0x214365).withOpacity(0.51);

      expect(instance1 == instance2, false);
    });
  });
}


lib/src/color.dart
import 'dart:math';

import 'package:color_wand/src/hsl_color.dart';
import 'package:color_wand/src/rgb_color.dart';
import 'package:meta/meta.dart';

/// Represents a color.
///
/// A [Color] can be constructed by specifying its value as either a Hex value,
/// a RGB vector, an RGBA (where A stands for alpha) vector,
/// a RGBO (where O stands for opacity) vector,
/// a HSL vector or a HSLO (where O stands for opacity) value.
///
/// Take a look at the underlying implementations ([RgbColor] and [HslColor])
/// for more info on how these values and vectors should look like.
@immutable
abstract class Color {
  static const maxOpacity = 1;

  const Color();

  const factory Color.hex(int value) = RgbColor.fromHex;

  const factory Color.rgb(
      {@required int red,
      @required int green,
      @required int blue}) = RgbColor.fromRgba;

  const factory Color.rgba(
      {@required int red,
      @required int green,
      @required int blue,
      int alpha}) = RgbColor.fromRgba;

  const factory Color.rgbo(
      {@required int red,
      @required int green,
      @required int blue,
      num opacity}) = RgbColor.fromRgbo;

  const factory Color.hsl(
      {@required num hue,
      @required num saturation,
      @required num lightness}) = HslColor.fromHslo;

  const factory Color.hslo(
      {@required num hue,
      @required num saturation,
      @required num lightness,
      num opacity}) = HslColor.fromHslo;

  static Color parseHex(String hex) {
    final withoutHash = hex.replaceAll('#', '');
    final value = int.parse(withoutHash, radix: 16);
    return RgbColor.fromHex(value);
  }

  num get opacity;

  bool get isOpaque => opacity == maxOpacity;

  bool get isTransparent => opacity != maxOpacity;

  /// The relative luminance of the color, as specified by
  /// https://www.w3.org/TR/2008/REC-WCAG20-20081211/#relativeluminancedef.
  ///
  /// Only valid for opaque colors; use [withBackground] first if necessary.
  num get _relativeLuminance {
    double _luminanceChannel(double v) =>
        (v <= 0.03928) ? v / 12.92 : pow((v + 0.055) / 1.055, 2.4);

    final asRgb = toRgb();
    return _luminanceChannel(asRgb.red / 255) * 0.2126 +
        _luminanceChannel(asRgb.green / 255) * 0.7152 +
        _luminanceChannel(asRgb.blue / 255) * 0.0722;
  }

  bool get isDark => _relativeLuminance <= 0.72;

  bool get isBright => _relativeLuminance > 0.72;

  // -----
  // Conversion
  // -----

  HslColor toHsl();

  RgbColor toRgb();

  // -----
  // Manipulation
  // -----

  /// Returns a new instance with [opacity].
  Color withOpacity(double opacity);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Color && this.toRgb().value == other.toRgb().value;

  @override
  int get hashCode => this.toRgb().hashCode;
}


lib/src/rgb_color.dart
import 'dart:math';

import 'package:color_wand/src/color.dart';
import 'package:color_wand/src/hsl_color.dart';
import 'package:meta/meta.dart';

/// A [RgbColor] stored as a 32 bit integer.
@immutable
class RgbColor extends Color {
  static const maxAlpha = 255;

  static const minAlpha = 0;

  final int value;

  /// Fractional representing the [opacity] of a color
  /// typically ranging between 0 and 1 inclusive.
  @override
  num get opacity => num.parse((alpha / maxAlpha).toStringAsFixed(2));

  /// Integer number representing the [alpha] value of a color
  /// ranging between 0 and 255 (inclusive).
  int get alpha => (value & 0xFF000000) >> 24;

  /// Integer number representing the [red] value of a color
  /// ranging between 0 and 255 (inclusive).
  int get red => (value & 0x00FF0000) >> 16;

  /// Integer number representing the [green] value of a color
  /// ranging between 0 and 255 (inclusive).
  int get green => (value & 0x0000FF00) >> 8;

  /// Integer number representing the [blue] value of a color
  /// ranging between 0 and 255 (inclusive).
  int get blue => (value & 0x000000FF) >> 0;

  // -----
  // Constructors
  // -----

  /// Construct a color from the lower 32 bits of an [int].
  ///
  /// The bits are interpreted as follows:
  ///
  /// * Bits 24-31 are the alpha value.
  /// * Bits 16-23 are the red value.
  /// * Bits 8-15 are the green value.
  /// * Bits 0-7 are the blue value.
  ///
  /// In other words, if AA is the alpha value in hex, RR the red value in hex,
  /// GG the green value in hex, and BB the blue value in hex, a color can be
  /// expressed as `const Color(0xAARRGGBB)`.
  const RgbColor(int value)
      : assert(value != null),
        value = value & 0xFFFFFFFF;

  /// Construct a color from the lower 32 bits of an [int].
  ///
  /// The bits are interpreted as follows:
  ///
  /// * Bits 24-31 are the red value.
  /// * Bits 16-23 are the green value.
  /// * Bits 8-15 are the blue value.
  /// * Bits 0-7 are the alpha value.
  ///
  /// In other words, if RR is the red value in hex, GG the green value in hex,
  /// BB the blue value in hex, and AA the alpha value in hex, a color can be
  /// expressed as `const Color(0xRRGGBBAA)`.
  ///
  /// If you omit the alpha value then the color will be fully opaque.
  // TODO: Maybe adjust how out of range values will get brought into range.
  const RgbColor.fromHex(int hex)
      : assert(hex != null),
        value = (((hex > 0xFFFFFF ? (hex >> 8) : hex) & 0xFFFFFF) |
                ((hex > 0xFFFFFF ? (hex & 0xFF) : 0xFF) << 24)) &
            0xFFFFFFFF;

  /// Construct a color from the lower 8 bits of four integers.
  ///
  /// * `r` is [red], from 0 to 255.
  /// * `g` is [green], from 0 to 255.
  /// * `b` is [blue], from 0 to 255.
  /// * `a` is the alpha value, with 0 being transparent and 255 being fully
  ///   opaque.
  ///
  /// Out of range values are brought into range using modulo 255.
  ///
  /// See also [fromRgbo], which takes the alpha value as a floating point
  /// value.
  const RgbColor.fromRgba(
      {@required int red, @required int green, @required int blue, int alpha})
      : assert(red != null),
        assert(green != null),
        assert(blue != null),
        value = (((alpha ?? maxAlpha) & 0xFF) << 24) |
            ((red & 0xFF) << 16) |
            ((green & 0xFF) << 8) |
            ((blue & 0xFF) << 0) & 0xFFFFFFFF;

  /// Construct a color from the lower 8 bits of four integers.
  ///
  /// * [red] ranges from 0 to 255.
  /// * [green] ranges from 0 to 255.
  /// * [blue] ranges from 0 to 255.
  /// * [opacity] with 0 being transparent and 1 being fully
  ///   opaque.
  ///
  /// Out of range [red], [green] and [blue] values are brought into range
  /// using modulo 255. [opacity] is brought into range using modulo 1.
  ///
  /// See also [fromRgba], which takes the opacity as an integer value.
  const RgbColor.fromRgbo(
      {@required int red, @required int green, @required int blue, num opacity})
      : assert(red != null),
        assert(green != null),
        assert(blue != null),
        value = (((maxAlpha + ((1 / (opacity ?? Color.maxOpacity)) ~/ 2)) ~/
                        (1 / (opacity ?? Color.maxOpacity)) &
                    0xff) <<
                24) |
            ((red & 0xFF) << 16) |
            ((green & 0xFF) << 8) |
            ((blue & 0xFF) << 0) & 0xFFFFFFFF;

  // -----
  // Manipulation
  // -----

  /// Returns a new instance with [alpha].
  Color withAlpha(int alpha) {
    assert(alpha != null);
    return _copy(alpha: alpha);
  }

  /// Returns a new instance with [opacity].
  @override
  Color withOpacity(num opacity) {
    assert(opacity != null);
    return _copy(opacity: opacity);
  }

  // -----
  // Conversion
  // -----

  /// Using https://en.wikipedia.org/wiki/HSL_and_HSV#RGB_to_HSL_and_HSV to
  /// convert.
  @override
  HslColor toHsl() {
    num rf = red / 255;
    num gf = green / 255;
    num bf = blue / 255;
    num cMax = [rf, gf, bf].reduce(max);
    num cMin = [rf, gf, bf].reduce(min);
    num delta = cMax - cMin;
    num hue;
    num saturation;
    num lightness;

    if (cMax == rf) {
      hue = 60 * ((gf - bf) / delta % 6);
    } else if (cMax == gf) {
      hue = 60 * ((bf - rf) / delta + 2);
    } else {
      hue = 60 * ((rf - gf) / delta + 4);
    }

    if (hue.isNaN || hue.isInfinite) {
      hue = 0;
    }

    lightness = (cMax + cMin) / 2;

    if (delta == 0) {
      saturation = 0;
    } else {
      saturation = delta / (1 - (lightness * 2 - 1).abs());
    }

    return HslColor.fromHslo(
        hue: hue,
        saturation: saturation * 100,
        lightness: lightness * 100,
        opacity: opacity);
  }

  @override
  RgbColor toRgb() => this;

  // -----
  // Util
  // -----

  RgbColor _copy({int red, int green, int blue, int alpha, num opacity}) =>
      opacity == null
          ? RgbColor.fromRgba(
              red: red ?? this.red,
              green: green ?? this.green,
              blue: blue ?? this.blue,
              alpha: alpha ?? this.alpha)
          : RgbColor.fromRgbo(
              red: red ?? this.red,
              green: green ?? this.green,
              blue: blue ?? this.blue,
              opacity: opacity ?? this.opacity);

  @override
  int get hashCode => value.hashCode;
}
